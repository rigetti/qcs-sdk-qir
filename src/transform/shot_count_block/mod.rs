mod pattern;
pub(crate) mod qir;
pub(crate) mod quil;

pub(crate) const PARAMETER_MEMORY_REGION_NAME: &str = "__qir_param";
