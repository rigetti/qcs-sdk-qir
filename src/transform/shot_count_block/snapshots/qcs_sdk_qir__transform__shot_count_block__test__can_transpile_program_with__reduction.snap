---
source: src/transform/shot_count_block/mod.rs
expression: context.module.print_to_string().to_str().unwrap()

---
; ModuleID = 'qcs'
source_filename = "./test/fixtures/programs/reduction.ll"

%Qubit = type opaque
%Result = type opaque
%Executable = type opaque
%ExecutionResult = type opaque

@parameter_memory_region_name = private unnamed_addr constant [12 x i8] c"__qir_param\00", align 1
@quil_program = private unnamed_addr constant [40 x i8] c"DECLARE ro BIT[1]\0AH 12\0AMEASURE 2 ro[0]\0A\00", align 1

declare void @__quantum__qis__h__body(%Qubit*) local_unnamed_addr

declare void @__quantum__qis__mz__body(%Qubit*, %Result*) local_unnamed_addr

declare i1 @__quantum__qis__read_result__body(%Result*) local_unnamed_addr

define internal fastcc void @QuantumApplication__Run__body() unnamed_addr {
entry:
  br label %body_execution

body:                                             ; preds = %body, %body_execution
  %0 = phi i64 [ %select, %body ], [ 0, %body_execution ]
  %1 = phi i64 [ %4, %body ], [ 1, %body_execution ]
  %2 = call i1 @get_readout_bit(%ExecutionResult* %7, i64 %1, i64 0)
  %3 = zext i1 %2 to i64
  %select = add i64 %0, %3
  %4 = add nuw nsw i64 %1, 1
  %5 = icmp ult i64 %1, 42
  br i1 %5, label %body, label %body_cleanup

body_cleanup:                                     ; preds = %body
  call void @free_executable(%Executable* %6)
  call void @free_execution_result(%ExecutionResult* %7)
  br label %exit

body_execution:                                   ; preds = %entry, <null operand!>
  %6 = call %Executable* @executable_from_quil(i8* getelementptr inbounds ([40 x i8], [40 x i8]* @quil_program, i32 0, i32 0))
  call void @wrap_in_shots(%Executable* %6, i32 42)
  %7 = call %ExecutionResult* @execute_on_qvm(%Executable* %6)
  call void @panic_on_failure(%ExecutionResult* %7)
  br label %body

exit:                                             ; preds = %body_cleanup, <null operand!>
  ret void
}

declare %Executable* @executable_from_quil(i8*)

declare %ExecutionResult* @execute_on_qpu(%Executable*, i8*)

declare %ExecutionResult* @execute_on_qvm(%Executable*)

declare void @free_executable(%Executable*)

declare void @free_execution_result(%ExecutionResult*)

declare i1 @get_readout_bit(%ExecutionResult*, i64, i64)

declare void @panic_on_failure(%ExecutionResult*)

declare void @set_param(%Executable*, i8*, i32, double)

declare void @wrap_in_shots(%Executable*, i32)

